### ISSUE TITLE
Maximum recursion depth exceeded when more than 30 nested functions encountered

### ISSUE THREAD

**Issue by iliyesku** [sqlfluff/sqlfluff#6259]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Linting the code, containing more than 30 nested functions throws an internal error.
When there are only 30 nested functions, everything is ok.

### Expected Behaviour

- The exception should be handled as a rule violation
- Maybe the limit should be increased

### Observed Behaviour

```
WARNING    Unable to lint my_model.sql due to an internal error. Please report this as an issue with your query's contents and stacktrace below!                                                                                                      
To hide this warning, add the failing file to .sqlfluffignore
Traceback (most recent call last):
  File ".venv\lib\site-packages\sqlfluff\core\linter\runner.py", line 110, in run
    yield partial()
  ... 2k lines
  File ".venv\lib\site-packages\sqlfluff\core\parser\grammar\base.py", line 214, in <genexpr>
    getattr(self, k, None) == getattr(other, k, None)
RecursionError: maximum recursion depth exceeded while calling a Python object
```

### How to reproduce

```sql
-- my_model.sql
SELECT
    REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(MY_COLUMN, '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', ''), '_', '') AS URL_URL
FROM MY_TABLE
```

```bash
$ sqlfluff lint my_model.sql
```

### Dialect

Snowflake

### Version

3.2.0

### Configuration

```
[sqlfluff]
templater = jinja
dialect = snowflake
```
or
```
[sqlfluff]
templater = dbt
dialect = snowflake
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
